In [33]:
import pandas as pd
import numpy as np
import os

# Raw - Mobility & Air quality

Importación de datos de calidad de aire y mobilidad

inicialmente se leen los datos crudos desde los archivos descargables de las  respectivas fuentes, se procede a compilar todos los datos de calidad de aire en 3 dataframe diferentes (Train, Test y Validation) que se convierten posteriormente a un archivo HDF5 el cual permite recolectar la información de manera más eficiente facilitando la lectura para proximas ocasiones, de igual maneraa se procede con los datos de mobilidad 

In [34]:
"""
directory_path = 'C:/Users/MARIA PAULA/Downloads/EAFIT/Proyecto de Grado/Air Quality/Train'
dfs = []

for filename in os.listdir(directory_path):
    if filename.endswith('.csv'):
        file_path = os.path.join(directory_path, filename)
        df = pd.read_csv(file_path)
        dfs.append(df)

Train_air = pd.concat(dfs, ignore_index=True)

# Save the combined DataFrame to an HDF5 file
Train_air.to_hdf('C:/Users/MARIA PAULA/Downloads/EAFIT/Proyecto de Grado/Air Quality/Train/train_air_dataframe.h5', key='df', mode='w') 
"""

"\ndirectory_path = 'C:/Users/MARIA PAULA/Downloads/EAFIT/Proyecto de Grado/Air Quality/Train'\ndfs = []\n\nfor filename in os.listdir(directory_path):\n    if filename.endswith('.csv'):\n        file_path = os.path.join(directory_path, filename)\n        df = pd.read_csv(file_path)\n        dfs.append(df)\n\nTrain_air = pd.concat(dfs, ignore_index=True)\n\n# Save the combined DataFrame to an HDF5 file\nTrain_air.to_hdf('C:/Users/MARIA PAULA/Downloads/EAFIT/Proyecto de Grado/Air Quality/Train/train_air_dataframe.h5', key='df', mode='w') \n"

In [35]:
# Load the DataFrame from the HDF5 file
Train_air = pd.read_hdf('C:/Users/MARIA PAULA/Downloads/EAFIT/Proyecto de Grado/Air Quality/Train/train_air_dataframe.h5', key='df')

Los datos de train van desde 2021-01-01 hasta 2023-05-01

In [36]:
Train_air.head()

Fecha_Hora  codigoSerial  pm25  calidad_pm25  pm10  calidad_pm10  \
0  2021-01-01 00:00:00            12  31.0           1.0  49.0           1.0   
1  2021-01-01 01:00:00            12  41.0           1.0  51.0           1.0   
2  2021-01-01 02:00:00            12  38.0           1.0  51.0           1.0   
3  2021-01-01 03:00:00            12  30.0           1.0  42.0           1.0   
4  2021-01-01 04:00:00            12  29.0           1.0  39.0           1.0   

      pm1  calidad_pm1      no  calidad_no  ...   p_ssr  calidad_p_ssr  \
0 -9999.0          1.0 -9999.0        4.31  ... -9999.0            1.0   
1 -9999.0          1.0 -9999.0        4.31  ... -9999.0            1.0   
2 -9999.0          1.0 -9999.0        4.31  ... -9999.0            1.0   
3 -9999.0          1.0 -9999.0        4.31  ... -9999.0            1.0   
4 -9999.0          1.0 -9999.0        4.31  ... -9999.0            1.0   

   pliquida_ssr  calidad_pliquida_ssr  rglobal_ssr  calidad_rglobal_ssr  \
0       -9999.0                   1.0      -9999.0                  1.0   
1       -9999.0                   1.0      -9999.0                  1.0   
2       -9999.0                   1.0      -9999.0                  1.0   
3       -9999.0                   1.0      -9999.0                  1.0   
4       -9999.0                   1.0      -9999.0                  1.0   

   taire10_ssr  calidad_taire10_ssr  vviento_ssr  calidad_vviento_ssr  
0      -9999.0                  1.0      -9999.0                  1.0  
1      -9999.0                  1.0      -9999.0                  1.0  
2      -9999.0                  1.0      -9999.0                  1.0  
3      -9999.0                  1.0      -9999.0                  1.0  
4      -9999.0                  1.0      -9999.0                  1.0  

[5 rows x 36 columns]

In [37]:
Train_air.columns

Index(['Fecha_Hora', 'codigoSerial', 'pm25', 'calidad_pm25', 'pm10',
       'calidad_pm10', 'pm1', 'calidad_pm1', 'no', 'calidad_no', 'no2',
       'calidad_no2', 'nox', 'calidad_nox', 'ozono', 'calidad_ozono', 'co',
       'calidad_co', 'so2', 'calidad_so2', 'pst', 'calidad_pst', 'dviento_ssr',
       'calidad_dviento_ssr', 'haire10_ssr', 'calidad_haire10_ssr', 'p_ssr',
       'calidad_p_ssr', 'pliquida_ssr', 'calidad_pliquida_ssr', 'rglobal_ssr',
       'calidad_rglobal_ssr', 'taire10_ssr', 'calidad_taire10_ssr',
       'vviento_ssr', 'calidad_vviento_ssr'],
      dtype='object')

In [38]:
Air = Train_air.copy()
estaciones_eliminar = [79,80,12] # se eliminan para tener la linea de tiempo de un solo sensor y no repetir fechas
Train_air = Train_air[~Train_air['codigoSerial'].isin(estaciones_eliminar)]

In [39]:
Air = Train_air[['Fecha_Hora', 'codigoSerial', 'pm25', 'calidad_pm25']]

Del dataset anterior únicamente se tomaran en cuenta las columnas de fecha_hora, codigoSerial, pm25 y calidad_pm25, ya que, las demás columnas al tener valores de -9999 y calidad 1 significan que los sensores están fuera de servicio y por la descripción del dataset se evidencia que solo las columnas de pm25 tienen un comportamiento diferente al mencionado

In [40]:
#Ponemos bajo un mismo formato a la columnda de Fecha_Hora
Air.loc[:,'Fecha_Hora'] = pd.to_datetime(Air['Fecha_Hora'], format='%Y-%m-%d %H:%M:%S')

In [41]:
# Establecer la columna 'fecha' como índice
Air.set_index('Fecha_Hora', inplace=True)

In [42]:
Air.head(10)

codigoSerial     pm25  calidad_pm25
Fecha_Hora                                              
2021-01-01 00:00:00            83  23.7741           1.0
2021-01-01 01:00:00            83  28.5095           1.0
2021-01-01 02:00:00            83  27.2400           1.0
2021-01-01 03:00:00            83  37.8129           1.0
2021-01-01 04:00:00            83  34.1848           1.0
2021-01-01 05:00:00            83  34.1002           1.0
2021-01-01 06:00:00            83  45.9056           1.0
2021-01-01 07:00:00            83  28.2544           1.0
2021-01-01 08:00:00            83  23.7000           1.0
2021-01-01 09:00:00            83  20.2378           1.0

La columna codigoSerial representa el código de la estación con esta podemos saber donde se encuentra ubicada, para poder comparar los datos con el dataset de mobilidad

Serial 83   
Comuna 16
Belén - I.E Pedro Justo Berrio

Serial 84
Comuna 14
El Poblado - I.E INEM Sede Santa Catalina

Serial 85
Comuna 10
CEN-TRAF - San Cristobal - Parque Biblioteca Fernando Botero

Serial 86
Comuna 4
Aranjuez - I.E Ciro Mendia

https://www.medellin.gov.co/es/conoce-algunos-datos-generales-de-la-ciudad/


Creamos una función para calcular el porcentaje de datos nulos en el dataset

In [43]:
def missing_values(df):
    Column_Names = df.columns

    Total_rows = df.shape[0]

    missing_values = df.isnull().sum()

    Percent_missing = (missing_values / Total_rows)*100
    
    result = pd.DataFrame({
        'Column_Name': Column_Names,
        'Missing_Values': missing_values.values,
        'Total_Rows': Total_rows,
        'Percent_Missing': Percent_missing.values
    })

    return result

Cambiamos la distribución de la tabla con ayuda de una tabla pivote, así por cada una de las estaciones tenemos una columna para calidad y otra para la magnitud de pm2.5

In [44]:
Air_organized = Air.pivot_table(index=Air.index, columns='codigoSerial', values=['pm25', 'calidad_pm25'], aggfunc = 'mean')

In [45]:
Air_organized.info()

<class 'pandas.core.frame.DataFrame'>
DatetimeIndex: 21144 entries, 2021-01-01 00:00:00 to 2023-05-31 23:00:00
Data columns (total 8 columns):
 #   Column              Non-Null Count  Dtype  
---  ------              --------------  -----  
 0   (calidad_pm25, 83)  21144 non-null  float64
 1   (calidad_pm25, 84)  21144 non-null  float64
 2   (calidad_pm25, 85)  21144 non-null  float64
 3   (calidad_pm25, 86)  21144 non-null  float64
 4   (pm25, 83)          21144 non-null  float64
 5   (pm25, 84)          21144 non-null  float64
 6   (pm25, 85)          21144 non-null  float64
 7   (pm25, 86)          21144 non-null  float64
dtypes: float64(8)
memory usage: 1.5 MB


In [46]:
missing_values(Air_organized)

Column_Name  Missing_Values  Total_Rows  Percent_Missing
0  (calidad_pm25, 83)               0       21144              0.0
1  (calidad_pm25, 84)               0       21144              0.0
2  (calidad_pm25, 85)               0       21144              0.0
3  (calidad_pm25, 86)               0       21144              0.0
4          (pm25, 83)               0       21144              0.0
5          (pm25, 84)               0       21144              0.0
6          (pm25, 85)               0       21144              0.0
7          (pm25, 86)               0       21144              0.0

Resulta que este dataset tiene la particularidad que los datos nulos o erroneos no son NaN, sino que se conocen por el valor de la calidad del datos, si es de1 el valor es real y ha sido revisado, si es un numero diferente, puede ser dudoso, malo, el sensor estar fuera de servicio, por lo que unicamente tomaremos los datos con valores de calidad de 2.6

In [47]:
# Revisamos la cantidad de datos con diferentes calificaciones de calidad
# Seleccionar todas las columnas con el nombre 'calidad_pm25' en el primer nivel del MultiIndex
calidad_pm25_columns = Air_organized.xs('calidad_pm25', level=0, axis=1)

# Iterar sobre cada columna (por código serial)
for column in calidad_pm25_columns.columns:
    # Contar la frecuencia de cada valor único
    value_counts = calidad_pm25_columns[column].value_counts()
    
    # Mostrar los resultados
    print(f"Columna (código serial): {column}")
    print(value_counts)
    print("\n")


Columna (código serial): 83
83
1.00    20113
2.30      256
4.31      197
3.10      160
4.32      107
2.20      102
3.30       87
2.40       66
3.00       52
3.20        4
Name: count, dtype: int64


Columna (código serial): 84
84
1.00    19967
2.30      204
4.31      179
3.00      175
3.10      172
2.40      144
3.30      120
2.20       92
4.32       85
3.20        6
Name: count, dtype: int64


Columna (código serial): 85
85
1.00    19622
4.31      335
2.30      279
3.10      237
4.32      235
3.30      152
2.20      104
3.00      102
2.40       75
3.20        3
Name: count, dtype: int64


Columna (código serial): 86
86
1.00    20060
2.30      322
4.31      201
3.10      172
3.30      147
4.32       94
2.20       66
3.00       53
2.40       23
3.20        6
Name: count, dtype: int64




In [48]:
Air_organized.describe()

calidad_pm25                                            \
codigoSerial            83            84            85            86   
count         21144.000000  21144.000000  21144.000000  21144.000000   
mean              1.104229      1.115980      1.167397      1.110002   
std               0.496410      0.509823      0.641769      0.507426   
min               1.000000      1.000000      1.000000      1.000000   
25%               1.000000      1.000000      1.000000      1.000000   
50%               1.000000      1.000000      1.000000      1.000000   
75%               1.000000      1.000000      1.000000      1.000000   
max               4.320000      4.320000      4.320000      4.320000   

                       pm25                                               
codigoSerial             83             84             85             86  
count          21144.000000   21144.000000   21144.000000   21144.000000  
mean             439.403983     366.326301     987.977519     397.001109  
std             7217.585234    6587.224924   10665.482414    6940.904490  
min            -9999.000000   -9999.000000   -9999.000000   -9999.000000  
25%               10.923350       9.590198       9.839388      12.116600  
50%               18.155250      14.647850      14.716250      17.618100  
75%               27.872825      20.360800      20.667425      23.770575  
max           117237.000000  117429.000000  100117.000000  108778.000000

Desde el diccionario de datos del SIATA menciona que valores de calidad mayores a 2.6 no son confiables por lo que se eliminan de mi dataset en aras de mejorar la calidad de mis datos y por ende la predicción

In [49]:
prueba = Air_organized.copy()

# Filtrar todas las columnas de 'calidad_pm25'
calidad_columns = prueba.xs('calidad_pm25', level=0, axis=1)

# Iterar sobre cada código serial en las columnas 'calidad_pm25'
for codigoSerial in calidad_columns:
    # Reemplazar los valores en 'pm25' si 'calidad_pm25' es menor a 2.6
    prueba.loc[prueba[('calidad_pm25', codigoSerial)] > 1, ('pm25', codigoSerial)] = np.nan
    #prueba.loc[prueba[('calidad_pm25', codigoSerial)] > 1, ('calidad_pm25', codigoSerial)] = np.nan


In [50]:
missing_values(prueba)

Column_Name  Missing_Values  Total_Rows  Percent_Missing
0  (calidad_pm25, 83)               0       21144         0.000000
1  (calidad_pm25, 84)               0       21144         0.000000
2  (calidad_pm25, 85)               0       21144         0.000000
3  (calidad_pm25, 86)               0       21144         0.000000
4          (pm25, 83)            1031       21144         4.876088
5          (pm25, 84)            1177       21144         5.566591
6          (pm25, 85)            1522       21144         7.198260
7          (pm25, 86)            1084       21144         5.126750

In [51]:
# Crear columnas auxiliares de año, mes, día a partir del índice
prueba['Año'] = prueba.index.year
prueba['Mes'] = prueba.index.month
prueba['Día'] = prueba.index.day

# Calcular la media horaria para cada día específico
# Esto incluye agrupar por año, mes, día  y luego calcular la media
medias_horarias = prueba.groupby(['Año', 'Mes', 'Día']).transform('mean')

# Rellenar valores faltantes usando las medias horarias calculadas
prueba.fillna(medias_horarias, inplace=True)

# Eliminar las columnas auxiliares si no son necesarias
prueba.drop(['Año', 'Mes', 'Día'], axis=1, inplace=True)

In [52]:
missing_values(prueba)

Column_Name  Missing_Values  Total_Rows  Percent_Missing
0  (calidad_pm25, 83)               0       21144         0.000000
1  (calidad_pm25, 84)               0       21144         0.000000
2  (calidad_pm25, 85)               0       21144         0.000000
3  (calidad_pm25, 86)               0       21144         0.000000
4          (pm25, 83)             144       21144         0.681044
5          (pm25, 84)              96       21144         0.454030
6          (pm25, 85)             216       21144         1.021566
7          (pm25, 86)              96       21144         0.454030

Al ya no necesitar los datos de calidad_pm25 pues hemos ya sacado/borrado las magnitudes asociadas a aquellas calidades diferentes de 1, sólo sacamos la descripción de las magnitudes por cada estación

In [53]:
prueba['pm25'].describe()

codigoSerial            83            84            85            86
count         21000.000000  21048.000000  20928.000000  21048.000000
mean             20.190187     15.606293     15.746424     18.770021
std              11.504186      8.070167      7.698368      8.714379
min               0.000000      0.000000      0.000000      0.000000
25%              11.204906      9.852787     10.160050     12.832100
50%              18.293750     14.700000     14.767300     17.677150
75%              27.334246     20.096025     19.937575     23.486575
max              78.207900     75.675600     65.921400    110.534000

In [54]:
missing_values(prueba['pm25'])

Column_Name  Missing_Values  Total_Rows  Percent_Missing
0          83             144       21144         0.681044
1          84              96       21144         0.454030
2          85             216       21144         1.021566
3          86              96       21144         0.454030

Para los datos nulos, se tienen 2 alternativas

1. **Forward or backward fill:** Use the value from the previous or next observation to fill null values. This method is especially useful for time series data.

df.fillna(method='ffill', inplace=True, axis=0)

2. **Interpolation:** Estimate missing values based on other values in the dataset. Pandas provides various interpolation methods such as linear, polynomial, or spline interpolation.

df.interpolate(method='linear', inplace=True, axis=0)

Inicialmente se llevara a cabo **Forward or backward fill** pues es más acorde al tipo de datos

In [55]:
# Forward fill null values
intento = prueba['pm25'].copy()
intento.ffill(axis=0, inplace=True)


In [56]:
intento.describe()

codigoSerial            83            84            85            86
count         21144.000000  21144.000000  21144.000000  21144.000000
mean             20.210572     15.614804     15.733463     18.766878
std              11.475611      8.053626      7.661464      8.694778
min               0.000000      0.000000      0.000000      0.000000
25%              11.276625      9.880947     10.216000     12.843075
50%              18.335950     14.727350     14.738050     17.677150
75%              27.352025     20.051225     19.877825     23.473657
max              78.207900     75.675600     65.921400    110.534000

| Rango de PM2.5 (µg/m³) | Nivel de Salud                     |
|--------------------------|------------------------------------|
| 0 - 50                   | Saludable                          |
| 51 - 100                 | Moderado                           |
| 101 - 150                | Poco saludable para grupos sensibles |
| 151 - 200                | Poco saludable para la población en general |
| Más de 200               | Muy poco saludable o peligroso     |


Para poder unir este dataset con el de mobilidad debo crear una nueva columna que tenga el numero de comuna al cual pertenece cada fila.

12 - comuna 10
79 - comuna 16 
80 - comuna 8
83 - comuna 16
84 - comuna 14
85 - comuna 10
86 - comuna 4

In [57]:
intento.describe()

codigoSerial            83            84            85            86
count         21144.000000  21144.000000  21144.000000  21144.000000
mean             20.210572     15.614804     15.733463     18.766878
std              11.475611      8.053626      7.661464      8.694778
min               0.000000      0.000000      0.000000      0.000000
25%              11.276625      9.880947     10.216000     12.843075
50%              18.335950     14.727350     14.738050     17.677150
75%              27.352025     20.051225     19.877825     23.473657
max              78.207900     75.675600     65.921400    110.534000

In [58]:
intento.columns.name = None
# Resetear el índice para tener 'Fecha_Hora' como columna
intento.reset_index(inplace=True)
intento.rename(columns={'index': 'Fecha_Hora'}, inplace=True)

# Mapeo de códigos seriales a comunas
mapa_comunas = {12: 10, 79: 16, 80: 8, 83: 16, 84: 14, 85: 10, 86: 4}

# Transformar el dataframe usando melt para desglosar por 'codigoSerial'
df_transformado = intento.melt(id_vars=['Fecha_Hora'], var_name='codigoSerial', value_name='pm_25')

# Crear una columna 'comuna' basada en el código serial
df_transformado['comuna'] = df_transformado['codigoSerial'].map(mapa_comunas)

# Resultado final
print(df_transformado)

               Fecha_Hora codigoSerial    pm_25  comuna
0     2021-01-01 00:00:00           83  23.7741      16
1     2021-01-01 01:00:00           83  28.5095      16
2     2021-01-01 02:00:00           83  27.2400      16
3     2021-01-01 03:00:00           83  37.8129      16
4     2021-01-01 04:00:00           83  34.1848      16
...                   ...          ...      ...     ...
84571 2023-05-31 19:00:00           86  21.2840       4
84572 2023-05-31 20:00:00           86  17.7081       4
84573 2023-05-31 21:00:00           86  15.3171       4
84574 2023-05-31 22:00:00           86  14.1202       4
84575 2023-05-31 23:00:00           86  12.9257       4

[84576 rows x 4 columns]


In [59]:
df_transformado.reset_index(drop=True, inplace=True)  # Esto elimina cualquier índice actual si existe

# Establecer 'Fecha_Hora' como el índice único
df_transformado.set_index('Fecha_Hora', inplace=True)

In [60]:
df_transformado

codigoSerial    pm_25  comuna
Fecha_Hora                                       
2021-01-01 00:00:00           83  23.7741      16
2021-01-01 01:00:00           83  28.5095      16
2021-01-01 02:00:00           83  27.2400      16
2021-01-01 03:00:00           83  37.8129      16
2021-01-01 04:00:00           83  34.1848      16
...                          ...      ...     ...
2023-05-31 19:00:00           86  21.2840       4
2023-05-31 20:00:00           86  17.7081       4
2023-05-31 21:00:00           86  15.3171       4
2023-05-31 22:00:00           86  14.1202       4
2023-05-31 23:00:00           86  12.9257       4

[84576 rows x 3 columns]

In [61]:
df_transformado.info()

<class 'pandas.core.frame.DataFrame'>
DatetimeIndex: 84576 entries, 2021-01-01 00:00:00 to 2023-05-31 23:00:00
Data columns (total 3 columns):
 #   Column        Non-Null Count  Dtype  
---  ------        --------------  -----  
 0   codigoSerial  84576 non-null  object 
 1   pm_25         84576 non-null  float64
 2   comuna        84576 non-null  int64  
dtypes: float64(1), int64(1), object(1)
memory usage: 2.6+ MB


In [62]:
df_final = df_transformado.drop(['codigoSerial'],axis =1)

In [63]:
df_final.to_hdf('C:/Users/MARIA PAULA/Downloads/EAFIT/Proyecto de Grado/Air Quality/Train/Train_set_1A.h5', key='df', mode='w', format='table')